File: Euler805.rs
use std::cmp::min;
use std::sync::Arc;
use std::thread;

const MOD: u64 = 1_000_000_007;

fn gcd(mut a: u64, mut b: u64) -> u64 {
    while b != 0 {
        let t = a % b;
        a = b;
        b = t;
    }
    a
}

fn extended_gcd(a: i64, b: i64) -> (i64, i64, i64) {
    if a == 0 {
        (b, 0, 1)
    } else {
        let (g, x1, y1) = extended_gcd(b % a, a);
        let x = y1 - (b / a) * x1;
        let y = x1;
        (g, x, y)
    }
}

fn mod_inverse(a: i64, m: i64) -> Option<i64> {
    let (g, x, _) = extended_gcd(a, m);
    if g != 1 {
        None
    } else {
        let mut res = x % m;
        if res < 0 {
            res += m;
        }
        Some(res)
    }
}

fn mod_pow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    if modulus == 1 {
        return 0;
    }
    let mut result: u64 = 1;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = ((result as u128 * base as u128) % modulus as u128) as u64;
        }
        base = ((base as u128 * base as u128) % modulus as u128) as u64;
        exp >>= 1;
    }
    result
}

struct PrimeFactorizer {
    spf: Vec<u32>,
}

impl PrimeFactorizer {
    fn new(max_n: usize) -> Self {
        let mut spf = vec![0u32; max_n + 1];
        for i in 0..=max_n {
            spf[i] = i as u32;
        }
        if max_n >= 1 {
            spf[1] = 1;
        }
        let limit = (max_n as f64).sqrt() as usize;
        for i in 2..=limit {
            if spf[i] == i as u32 {
                for j in (i * i..=max_n).step_by(i) {
                    if spf[j] == j as u32 {
                        spf[j] = i as u32;
                    }
                }
            }
        }
        Self { spf }
    }

    fn get_factors(&self, mut n: u32) -> Vec<u32> {
        let mut factors = Vec::new();
        while n > 1 {
            let p = self.spf[n as usize];
            factors.push(p);
            n /= p;
        }
        factors
    }

    fn euler_phi(&self, n: u32) -> u32 {
        let mut result = n;
        let mut temp = n;
        while temp > 1 {
            let p = self.spf[temp as usize];
            while temp % p == 0 {
                temp /= p;
            }
            result -= result / p;
        }
        result
    }
}

fn multiplicative_order(a: u64, m: u64, factorizer: &PrimeFactorizer) -> Option<u64> {
    if gcd(a, m) != 1 {
        return None;
    }
    if m == 1 {
        return Some(1);
    }
    let phi = factorizer.euler_phi(m as u32) as u64;
    let mut factors = factorizer.get_factors(phi as u32);
    factors.sort_unstable();
    factors.dedup();

    let mut order = phi;
    for &p in &factors {
        let p_u64 = p as u64;
        while order % p_u64 == 0 && mod_pow(a, order / p_u64, m) == 1 {
            order /= p_u64;
        }
    }
    Some(order)
}

fn min_l_magnitude(p: u64, q: u64) -> u64 {
    let limit = (q + p - 1) / p;
    let mut pow10 = 1u64;
    let mut l = 1u64;
    while pow10 < limit {
        pow10 *= 10;
        l += 1;
    }
    l
}

fn solve_n(u: u64, v: u64, factorizer: &PrimeFactorizer) -> u64 {
    let p = u * u * u;
    let q = v * v * v;
    if p >= 10 * q {
        return 0;
    }

    let k_val = 10 * q - p;
    let min_l = min_l_magnitude(p, q);

    let mut best_l = u64::MAX;
    let mut best_d = 10u64;

    for d in 1..=9u64 {
        if p * (d + 1) >= 10 * q {
            continue;
        }

        let g = gcd(k_val, d);
        let m_prime = k_val / g;
        if gcd(10, m_prime) != 1 {
            continue;
        }

        let l_base = match multiplicative_order(10, m_prime, factorizer) {
            Some(ord) => ord,
            None => continue,
        };

        let l_cand = ((min_l + l_base - 1) / l_base) * l_base;
        if l_cand < best_l || (l_cand == best_l && d < best_d) {
            best_l = l_cand;
            best_d = d;
        }
    }

    if best_l == u64::MAX {
        return 0;
    }

    let ten_pow_l = mod_pow(10, best_l, MOD);
    let term1 = ((best_d % MOD) as u128 * (q % MOD) as u128 % MOD as u128) as u64;
    let term2 = (ten_pow_l + MOD - 1) % MOD;
    let k_inv = mod_inverse(k_val as i64, MOD as i64).unwrap() as u64;

    let mut n_mod = (term1 as u128 * term2 as u128 % MOD as u128) as u64;
    n_mod = (n_mod as u128 * k_inv as u128 % MOD as u128) as u64;
    n_mod
}

fn compute_total_serial(m: u64, factorizer: &PrimeFactorizer) -> u64 {
    let mut total = 0u64;
    for u in 1..=m {
        for v in 1..=m {
            if gcd(u, v) == 1 {
                total += solve_n(u, v, factorizer);
                if total >= MOD {
                    total %= MOD;
                }
            }
        }
    }
    total % MOD
}

fn compute_total(m: u64, factorizer: &PrimeFactorizer, threads: usize) -> u64 {
    if threads <= 1 {
        return compute_total_serial(m, factorizer);
    }

    let m_usize = m as usize;
    let threads = min(threads, m_usize.max(1));
    let chunk = (m_usize + threads - 1) / threads;
    let factorizer = Arc::new(factorizer);

    let mut handles = Vec::new();
    for t in 0..threads {
        let start = t * chunk + 1;
        if start > m_usize {
            break;
        }
        let end = min(m_usize, (t + 1) * chunk);
        let factorizer = Arc::clone(&factorizer);
        handles.push(thread::spawn(move || {
            let mut local = 0u64;
            for u in start..=end {
                let u64u = u as u64;
                for v in 1..=m {
                    if gcd(u64u, v) == 1 {
                        local += solve_n(u64u, v, factorizer.as_ref());
                        if local >= MOD {
                            local %= MOD;
                        }
                    }
                }
            }
            local % MOD
        }));
    }

    let mut total = 0u64;
    for handle in handles {
        total = (total + handle.join().unwrap()) % MOD;
    }
    total
}

fn run_validations(factorizer: &PrimeFactorizer) -> bool {
    let mut ok = true;

    let t1 = compute_total_serial(1, factorizer);
    if t1 != 1 {
        eprintln!("Validation failed: T(1) expected 1, got {}", t1);
        ok = false;
    } else {
        eprintln!("Validation checkpoint: T(1) ok.");
    }

    let expected_t3 = 262429173u64;
    let t3 = compute_total_serial(3, factorizer);
    if t3 != expected_t3 {
        eprintln!(
            "Validation failed: T(3) expected {}, got {}",
            expected_t3, t3
        );
        ok = false;
    } else {
        eprintln!("Validation checkpoint: T(3) ok.");
    }

    if ok {
        eprintln!("All validation checkpoints passed.");
    }
    ok
}

fn main() {
    let m_limit: u64 = 200;
    let max_n = 80_000_000usize; // 10 * 200^3

    eprintln!("Building SPF up to {}...", max_n);
    let factorizer = PrimeFactorizer::new(max_n);

    eprintln!("Running validation checkpoints...");
    if !run_validations(&factorizer) {
        return;
    }

    let threads = thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(1)
        .min(16);

    let total = compute_total(m_limit, &factorizer, threads);
    println!("T({}) modulo {}: {}", m_limit, MOD, total);
}

File: Euler805.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <limits>
#include <string>
#include <thread>
#include <vector>

namespace {

using u64 = std::uint64_t;
using i64 = std::int64_t;
using u128 = __uint128_t;

constexpr u64 MOD = 1'000'000'007ULL;

struct Options {
    int m = 200;
    int threads = 0;  // 0 => auto
    bool run_checkpoints = true;
};

bool parse_int_after_prefix(const std::string& arg, const std::string& prefix, int& value) {
    if (arg.rfind(prefix, 0U) != 0U) {
        return false;
    }
    const std::string tail = arg.substr(prefix.size());
    if (tail.empty()) {
        return false;
    }
    int parsed = 0;
    for (char ch : tail) {
        if (ch < '0' || ch > '9') {
            return false;
        }
        parsed = parsed * 10 + static_cast<int>(ch - '0');
    }
    value = parsed;
    return true;
}

bool parse_arguments(int argc, char** argv, Options& options) {
    for (int i = 1; i < argc; ++i) {
        const std::string arg(argv[i]);
        if (arg == "--skip-checkpoints") {
            options.run_checkpoints = false;
            continue;
        }
        if (parse_int_after_prefix(arg, "--m=", options.m) ||
            parse_int_after_prefix(arg, "--threads=", options.threads)) {
            continue;
        }
        std::cerr << "Unknown argument: " << arg << '\n';
        return false;
    }
    return options.m >= 1 && options.threads >= 0;
}

u64 gcd_u64(u64 a, u64 b) {
    while (b != 0ULL) {
        const u64 t = a % b;
        a = b;
        b = t;
    }
    return a;
}

u64 mod_pow(u64 base, u64 exp, const u64 mod) {
    u64 result = 1ULL;
    base %= mod;
    while (exp > 0ULL) {
        if (exp & 1ULL) {
            result = static_cast<u64>((static_cast<u128>(result) * base) % mod);
        }
        base = static_cast<u64>((static_cast<u128>(base) * base) % mod);
        exp >>= 1ULL;
    }
    return result;
}

i64 extended_gcd(i64 a, i64 b, i64& x, i64& y) {
    if (a == 0) {
        x = 0;
        y = 1;
        return b;
    }
    i64 x1 = 0;
    i64 y1 = 0;
    const i64 g = extended_gcd(b % a, a, x1, y1);
    x = y1 - (b / a) * x1;
    y = x1;
    return g;
}

u64 mod_inverse(u64 a, const u64 mod) {
    i64 x = 0;
    i64 y = 0;
    const i64 g = extended_gcd(static_cast<i64>(a), static_cast<i64>(mod), x, y);
    if (g != 1) {
        return 0ULL;
    }
    i64 v = x % static_cast<i64>(mod);
    if (v < 0) {
        v += static_cast<i64>(mod);
    }
    return static_cast<u64>(v);
}

class PrimeFactorizer {
public:
    explicit PrimeFactorizer(const int max_n) {
        spf_.resize(static_cast<std::size_t>(max_n + 1));
        for (int i = 0; i <= max_n; ++i) {
            spf_[static_cast<std::size_t>(i)] = static_cast<std::uint32_t>(i);
        }
        if (max_n >= 1) {
            spf_[1] = 1U;
        }
        for (int i = 2; i <= max_n / i; ++i) {
            if (spf_[static_cast<std::size_t>(i)] != static_cast<std::uint32_t>(i)) {
                continue;
            }
            for (int j = i * i; j <= max_n; j += i) {
                if (spf_[static_cast<std::size_t>(j)] == static_cast<std::uint32_t>(j)) {
                    spf_[static_cast<std::size_t>(j)] = static_cast<std::uint32_t>(i);
                }
            }
        }
    }

    u64 euler_phi(u64 n) const {
        u64 result = n;
        u64 x = n;
        while (x > 1ULL) {
            const u64 p = static_cast<u64>(spf_[static_cast<std::size_t>(x)]);
            while (x % p == 0ULL) {
                x /= p;
            }
            result -= result / p;
        }
        return result;
    }

    std::vector<u64> prime_factors_with_multiplicity(u64 n) const {
        std::vector<u64> factors;
        while (n > 1ULL) {
            const u64 p = static_cast<u64>(spf_[static_cast<std::size_t>(n)]);
            factors.push_back(p);
            n /= p;
        }
        return factors;
    }

private:
    std::vector<std::uint32_t> spf_;
};

u64 multiplicative_order(const u64 a, const u64 m, const PrimeFactorizer& factorizer) {
    if (gcd_u64(a, m) != 1ULL) {
        return 0ULL;
    }
    if (m == 1ULL) {
        return 1ULL;
    }

    u64 order = factorizer.euler_phi(m);
    std::vector<u64> factors = factorizer.prime_factors_with_multiplicity(order);
    std::sort(factors.begin(), factors.end());
    factors.erase(std::unique(factors.begin(), factors.end()), factors.end());

    for (const u64 p : factors) {
        while (order % p == 0ULL && mod_pow(a, order / p, m) == 1ULL) {
            order /= p;
        }
    }
    return order;
}

u64 min_l_magnitude(const u64 p, const u64 q) {
    const u64 limit = (q + p - 1ULL) / p;
    u64 pow10 = 1ULL;
    u64 l = 1ULL;
    while (pow10 < limit) {
        pow10 *= 10ULL;
        ++l;
    }
    return l;
}

u64 solve_pair(const u64 u, const u64 v, const PrimeFactorizer& factorizer) {
    const u64 p = u * u * u;
    const u64 q = v * v * v;
    if (p >= 10ULL * q) {
        return 0ULL;
    }

    const u64 k_val = 10ULL * q - p;
    const u64 min_l = min_l_magnitude(p, q);
    u64 best_l = std::numeric_limits<u64>::max();
    u64 best_d = 10ULL;

    for (u64 d = 1ULL; d <= 9ULL; ++d) {
        if (p * (d + 1ULL) >= 10ULL * q) {
            continue;
        }
        const u64 g = gcd_u64(k_val, d);
        const u64 m_prime = k_val / g;
        if (gcd_u64(10ULL, m_prime) != 1ULL) {
            continue;
        }
        const u64 l_base = multiplicative_order(10ULL, m_prime, factorizer);
        if (l_base == 0ULL) {
            continue;
        }
        const u64 l_cand = ((min_l + l_base - 1ULL) / l_base) * l_base;
        if (l_cand < best_l || (l_cand == best_l && d < best_d)) {
            best_l = l_cand;
            best_d = d;
        }
    }

    if (best_l == std::numeric_limits<u64>::max()) {
        return 0ULL;
    }

    const u64 ten_pow_l = mod_pow(10ULL, best_l, MOD);
    const u64 term1 = static_cast<u64>((static_cast<u128>(best_d % MOD) * (q % MOD)) % MOD);
    const u64 term2 = (ten_pow_l + MOD - 1ULL) % MOD;
    const u64 k_inv = mod_inverse(k_val % MOD, MOD);
    u64 out = static_cast<u64>((static_cast<u128>(term1) * term2) % MOD);
    out = static_cast<u64>((static_cast<u128>(out) * k_inv) % MOD);
    return out;
}

u64 compute_total_serial(const int m, const PrimeFactorizer& factorizer) {
    u64 total = 0ULL;
    for (int u = 1; u <= m; ++u) {
        const u64 uu = static_cast<u64>(u);
        for (int v = 1; v <= m; ++v) {
            if (gcd_u64(uu, static_cast<u64>(v)) == 1ULL) {
                total += solve_pair(uu, static_cast<u64>(v), factorizer);
                if (total >= MOD) {
                    total %= MOD;
                }
            }
        }
    }
    return total % MOD;
}

u64 compute_total_parallel(const int m, const PrimeFactorizer& factorizer, int threads) {
    if (threads <= 1) {
        return compute_total_serial(m, factorizer);
    }
    threads = std::min(threads, std::max(1, m));
    const int chunk = (m + threads - 1) / threads;
    std::vector<std::thread> workers;
    std::vector<u64> partial(static_cast<std::size_t>(threads), 0ULL);

    for (int t = 0; t < threads; ++t) {
        const int start = t * chunk + 1;
        if (start > m) {
            break;
        }
        const int end = std::min(m, (t + 1) * chunk);
        workers.emplace_back([&, t, start, end]() {
            u64 local = 0ULL;
            for (int u = start; u <= end; ++u) {
                const u64 uu = static_cast<u64>(u);
                for (int v = 1; v <= m; ++v) {
                    if (gcd_u64(uu, static_cast<u64>(v)) == 1ULL) {
                        local += solve_pair(uu, static_cast<u64>(v), factorizer);
                        if (local >= MOD) {
                            local %= MOD;
                        }
                    }
                }
            }
            partial[static_cast<std::size_t>(t)] = local % MOD;
        });
    }
    for (std::thread& th : workers) {
        th.join();
    }
    u64 total = 0ULL;
    for (u64 v : partial) {
        total += v;
        if (total >= MOD) {
            total %= MOD;
        }
    }
    return total % MOD;
}

bool run_checkpoints(const PrimeFactorizer& factorizer, const int threads) {
    if (compute_total_serial(1, factorizer) != 1ULL) {
        std::cerr << "Checkpoint failed: T(1)=1" << '\n';
        return false;
    }
    if (compute_total_serial(3, factorizer) != 262'429'173ULL) {
        std::cerr << "Checkpoint failed: T(3)=262429173" << '\n';
        return false;
    }
    if (threads > 1) {
        const u64 serial = compute_total_serial(6, factorizer);
        const u64 parallel = compute_total_parallel(6, factorizer, threads);
        if (serial != parallel) {
            std::cerr << "Checkpoint failed: deterministic serial/parallel equality" << '\n';
            return false;
        }
    }
    return true;
}

}  // namespace

int main(int argc, char** argv) {
    Options options;
    if (!parse_arguments(argc, argv, options)) {
        return 1;
    }

    const int threads = (options.threads > 0)
                            ? options.threads
                            : static_cast<int>(std::max(1U, std::thread::hardware_concurrency()));
    const u64 max_n = 10ULL * static_cast<u64>(options.m) * static_cast<u64>(options.m) *
                      static_cast<u64>(options.m);
    if (max_n > static_cast<u64>(std::numeric_limits<int>::max())) {
        std::cerr << "Requested m is too large for SPF table bounds" << '\n';
        return 1;
    }

    PrimeFactorizer factorizer(static_cast<int>(max_n));
    if (options.run_checkpoints && !run_checkpoints(factorizer, threads)) {
        return 2;
    }

    std::cout << compute_total_parallel(options.m, factorizer, threads) << '\n';
    return 0;
}
